parametric/test_span_sampling.py

```diff
@@ -204,27 +204,27 @@ def test_multi_rule_drop_keep_span_sampling_sss007(test_agent, test_library):
 )
 def test_single_rule_rate_limiter_span_sampling_sss008(test_agent, test_library):
     """Test span sampling tags are added until rate limit hit, then need to wait for tokens to reset"""
-    # generate spans until we hit the rate limit
-    while True:
-        with test_library:
+    # generate three traces before requesting them to avoid timing issues
```

Comment:
The problem with current tests 008 and 013 is that they assume that wait_for_num_traces will complete in sub seconds but it's not always the case leading to the test failure. I added a print current timestamp to demonstrate when it fails. Here is the output:

```
Start Span in-loop at:  2022-12-21 13:23:40.808084
<parametric.conftest._TestSpan object at 0x10694bfa0>
Start Span in-loop at:  2022-12-21 13:23:40.850864
<parametric.conftest._TestSpan object at 0x10694b3d0>
Start Span in-loop at:  2022-12-21 13:23:40.883806
<parametric.conftest._TestSpan object at 0x10694b190>
Start Span after-loop at:  2022-12-21 13:23:42.068752
```

As we can see 4th span has a time span that is more than a second, from the previous span and the test fails because it assumes that the 4th span has been send within a second but it hasn't. In order to avoid this issue I've changed tests 008 and 013 to send bunch of spans together and then wait and analyze the result for all the traces knowing that they were created within one second time window.

This change also helped to catch a bug in the Python Tracer.

Reply:
thanks for the detailed write-up @ygree!